src/lib/core/backtrace.cc

```diff
@@ -375,97 +211,144 @@ __asm__ volatile(
 	"\tldp d12, d13, [x0, #16 * 8]\n"
 	"\tldp d14, d15, [x0, #16 * 9]\n"
 	"\tadd sp, x0, #8 * 20\n"
-	:
-	: "r" (unw_context), "r" (coro_ctx), "S" (unw_getcontext_f)
-	: /*"lr", "r0", "r1", "ip" */
-	 "x0", "x1", "x2", "x30"
-	);
-#endif
+	: "=m" (*bt)
+	: "m" (bt), "r" (&fiber->ctx), "S" (collect_current_stack)
+	: "x0", "x1", "x2", "x3", "x4", "x5", "x6", "x7", "x8", "x9", "x10",
+	  "x11", "x12", "x13", "x14", "x15", "x16", "x17", "x18",
+	  "s0", "s1", "s2", "s3", "s4", "s5", "s6", "s7", "s8", "s9", "s10",
+	  "s11", "s12", "s13", "s14", "s15", "s16", "s17", "s18", "s19", "s20",
+	  "s21", "s22", "s23", "s24", "s25", "s26", "s27", "s28", "s29", "s30",
+	  "s31", "d0", "d1", "d2", "d3", "d4", "d5", "d6", "d7");
+#endif /* __x86_64__ */
+skip_frames:
+	bt->frame_count -= skip_frames;
+	memmove(bt->frames, bt->frames + skip_frames,
+		sizeof(bt->frames[0]) * bt->frame_count);
 }
 
-/**
- * Call `cb' callback for each `coro_ctx' contained frame or the current
- * executed coroutine if `coro_ctx' is NULL. A coro_context is a structure
- * created on each coroutine yield to store execution context so for an
- * on-CPU coroutine there is no valid coro_context could be defined and
- * NULL is passed.
+/*
+ * Resolve function name and offset based on `ip`.
  */
-void
-backtrace_foreach(backtrace_cb cb, coro_context *coro_ctx, void *cb_ctx)
+static const char *
+core_backtrace_resolve_frame(unw_word_t ip, unw_word_t *offset,
+			     char **demangle_buf, size_t *demangle_buf_len)
 {
-	unw_cursor_t unw_cur;
-	unw_context_t unw_ctx_bt;
-	if (coro_ctx == NULL) {
-		/*
-		 * Current executing coroutine and simple unw_getcontext
-		 * should function.
-		 */
-		unw_getcontext(&unw_ctx_bt);
-	} else {
-		/*
-		 * Execution context is stored in the coro_ctx
-		 * so use special context-switching handler to
-		 * capture an unwind context.
-		 */
-		coro_unwcontext(&unw_ctx_bt, coro_ctx);
+	assert(offset != nullptr);
+	assert(demangle_buf != nullptr);
+	assert(demangle_buf_len != nullptr);
```

Comment:
I wouldn't use `nullptr`, as well as lambdas, auto keyword, aggregate initialization, and other C++-specific syntactic sugaring, because it doesn't simplify the code or make it easier to read, but only makes it incompatible with C, which is still the main language in Tarantool. In general, IMO it looks ugly when the code is written conceptually in C (i.e. with structs + functions), but uses some features from C++ for no real reason. AFAICS the only reason why `backtrace.cc` is a C++ file is the need to use `abi::__cxa_demangle`. I'd add a C wrapper around this function and convert `backtrace.cc` to C.

Reply:
Fixed: 
* wrapped `abi::__cxa_demangle` into `cxx_abi.{cc, h}`
* changed `backtrace.cc` to `backtrace.c`